### ISSUE TITLE
Support retaining file capabilities in COPY/ADD

### ISSUE THREAD

**Issue by mrunalp** [docker/docker#20435]
When one tries to copy in a binary in a Dockerfile that has some file capabilities set, then those capabilities are lost during the copy. This requires the Dockerfile author to RUN setcap </path/to/binary> and doing that results in a copy up of the file essentially doubling the storage required to store the binary. This can be avoided if the file capabilities are applied during COPY/ADD saving disk space.

**Comment by thaJeztah** [comment]
ping @estesp I think this is what we talked about

**Comment by thaJeztah** [comment]
@duglin something to consider for the `ADD` / `COPY` `--user` PR? (As a follow up)

**Comment by estesp** [comment]
@thaJeztah yes--so this is additional scope to just preserving ownership; in this case basically preserving all attributes of the original file

**Comment by thaJeztah** [comment]
Talking about this some more now; possibly it's a bug that these attributes are stripped, but we need to look into this

/cc @tonistiigi @dmcgowan

**Comment by schnatterer** [comment]
Is there any news on this?

I wonder why this issue hasn't seen more activity. I presume it should be fairly common for multi-stage builds targeting at a non-root image.

An example: I want to run my java web server on port 443, but don't want to run the container as root.
So what I need to do is basically:

```Dockerfile
USER root
RUN apt-get update && apt-get install -y libcap2-bin && \
   setcap CAP_NET_BIND_SERVICE=+ep $(which java)
USER whatever
```

When I do this in a early stage and `COPY` the `java` binary to the final stage, it does no longer have the desired capability (`getcap $(which java)`). 
Another example would be `ping`.

So currently I either have to
* run the steps above in the final image, which bloats it or
* need to setup `authbind`, which is not the least privilege approach here.

Both of which are not ideal. Any better workarounds here?

**Comment by tonistiigi** [comment]
@schnatterer This works with buildkit enabled.

```
from debian AS base
run apt-get update && apt-get install -y libcap2-bin

from base as mid
run echo aa > /foo && setcap CAP_NET_BIND_SERVICE=+ep /foo

from base
copy --from=mid /foo /bar
run getcap /bar
```

`DOCKER_BUILDKIT=1 docker build --progress=plain .`

```
...
#8 [stage-2 2/2] RUN getcap /bar
#8 0.757 /bar = cap_net_bind_service+ep
#8 DONE 0.9s
...
```

**Comment by schnatterer** [comment]
@tonistiigi Thanks for pointing this out this workaround. 
Same `Dockerfile` with the classical `docker build` results in the capability not being copied.

Slightly off-topic:
I normally use buildkit by default, but in the [project](https://github.com/schnatterer/smeagol-galore/blob/112b8bcd313cd3c375ba523c50a008bf144db6ef/Dockerfile) I encountered the capability  issue, buildkit (packaged with docker 19.03.8) fails with `cannot copy to non-directory`. The classic Docker Build succeeds here. 
Haven't had time to investigate, though.

**Comment by schnatterer** [comment]
My workaround at the moment looks as follows:

In an early stage:
```Dockerfile
FROM something as aggregator
RUN apt-get update
RUN cd /tmp && apt-get download authbind
RUN dpkg-deb -X /tmp/*.deb /dist
RUN touch /dist/etc/authbind/byport/443 /dist/etc/authbind/byport/80 && \
    chown 1001:0 /dist/etc/authbind/byport/* && \
    chmod 550 /dist/etc/authbind/byport/*

FROM something
# Don't --chown here, or some authbind will not work
COPY --from=aggregator /dist /
```

This does not impact my final image too much, as I copy the `/dist` folder anyway.
Complete example [here](https://github.com/schnatterer/smeagol-galore/blob/f78f3ef20b34b9331da55cf97e66c635d3823d1a/Dockerfile#L113).
